aptos-move/framework/aptos-framework/sources/fungible_asset/fungible_store.move

```diff
@@ -0,0 +1,314 @@
+                index: smart_table::new()
+            })
+        }
+    }
+
+    /// Get the `FungibleAssetWallet` object of `metadata` belonging to `account`.
+    /// if `create_on_demand` is true, an default `FungibleAssetWallet` will be created if not exist; otherwise abort.
+    fun get_account_fungible_asset_object(
+        account: address,
+        metadata: &Object<FungibleAssetMetadata>,
+        create_on_demand: bool
+    ): Option<Object<FungibleAssetWallet>> acquires FungibleAssetStore {
+        ensure_fungible_asset_store(account);
+        let metadata = fungible_asset::verify(metadata);
+        let index_table = &mut borrow_global_mut<FungibleAssetStore>(account).index;
+        if (!smart_table::contains(index_table, copy metadata)) {
+            if (create_on_demand) {
+                let afa_obj = create_account_fungible_asset_object(account, &metadata);
+                smart_table::add(index_table, copy metadata, afa_obj);
+            } else {
+                return option::none()
+            }
+        };
+        let wallet = *smart_table::borrow(index_table, metadata);
+        option::some(wallet)
+    }
+
+    /// Ensure the existence and return the `FungibleAssetWallet`.
+    inline fun ensure_fungible_asset_wallet(
+        account: address,
+        metadata: &Object<FungibleAssetMetadata>,
+        create_on_demand: bool
+    ): Object<FungibleAssetWallet> acquires FungibleAssetStore {
+        let afa_opt = get_account_fungible_asset_object(account, metadata, create_on_demand);
+        assert!(option::is_some(&afa_opt), error::not_found(EFUNGIBLE_ASSET_WALLET_OBJECT));
+        option::destroy_some(afa_opt)
+    }
+
+    /// Create a default `FungibleAssetWallet` object with zero balance of `metadata`.
+    fun create_account_fungible_asset_object(
+        account: address,
+        metadata: &Object<FungibleAssetMetadata>
+    ): Object<FungibleAssetWallet> {
+        // Must review carefully here.
+        let asset_signer = create_signer::create_signer(object::object_address(metadata));
+        let creator_ref = object::create_object_from_object(&asset_signer);
+        let wallet = fungible_asset::new_fungible_asset_wallet_object(&creator_ref, metadata);
+        // Transfer the owner to `account`.
+        object::transfer(&asset_signer, wallet, account);
+        // Disable transfer of coin object so the object itself never gets transfered.
+        let transfer_ref = object::generate_transfer_ref(&creator_ref);
+        object::disable_ungated_transfer(&transfer_ref);
+        wallet
+    }
+
+    /// Remove the `FungibleAssetWallet` object of `metadata` from `account` if balance drops to 0 and
+    /// `allowed_ungated_transfer` is allowed.
+    inline fun maybe_delete(wallet: Object<FungibleAssetWallet>) acquires FungibleAssetStore {
+        if (fungible_asset::balance(&wallet) == 0 && fungible_asset::ungated_transfer_allowed(&wallet)) {
```

Comment:
let's revisit the goal here, I don't think we just want to delete something if it is empty
consider the case where it is your primary wallet, I think that lifetime should be managed by the fungible_store

Reply:
This is something I am not quite sure. I can keep it for now, but also keep in mind that you may not want the wallets with 0 balance of different metadata objects that you just used once.